doc/SAI_Proposal_Generic_Extensions.md

```diff
@@ -0,0 +1,423 @@
+```json
+"attributes": [
+    {
+        "virtual_ip": {
+            "sai_metadata": {
+                "sai_attr_value_type": "SAI_ATTR_VALUE_TYPE_IP_ADDRESS"
+            },
+            "value": "10.1.1.1"
+        }
+    },
+    {
+        "nexthop_list": {
+            "sai_metadata": {
+                "sai_attr_value_type": "SAI_ATTR_VALUE_TYPE_OBJECT_LIST"
+            },
+            "value": [ "0x1500000000001fa", "0x1500000000001fb" ]
+        }
+    }
+]
+```
+
+## SAI ACL considerations ##
+The new VIP table is conditionally evaluated based on policies in the ACL tables. This requires a new action type in saiacl.h.
+SAI ACL action type is today exposed via 2 mechanisms. * sai_acl_action_type_t - This type is only used today to query the SAI ACL action capabilities * SAI_ACL_ENTRY_ATTR_ACTION_START/END - An attribute specifiying the ACL action type for an entry
+While there are multiple mechanisms to extend these, one recommendation is as below. 
+- Extend sai_acl_acion_type_t with a custom range just like all the attribute enums for SAI objects 
+  - SAI_ACL_ACTION_TYPE_CUSTOM_RANGE 
+  - SAI_ACL_ACTION_TYPE_DO_LB = SAI_ACL_ACTION_TYPE_CUSTOM_RANGE + 1 
+- Add a new custom attribute. This attribute essentially acts as an input into the new VIP table filtering packets. 
+```c
+    /**
+     * @brief Apply vip table
+     *
+     * @type boolean
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+   SAI_ACL_ENTRY_ACTION_DO_LB = SAI_ACL_ENTRY_ACTION_CUSTOM_RANGE + 1
+```
+
+## Why is metadata required? ##
+1. Data validation for intermediate SW layers. sonic-sairedis in the SONiC operating system is an example. The meta infra can understand the object types from the JSON data type and perform the relevant validations.
+2. Reference counting. The OID on its own can just be a sequentially incrementing integer and does not hold any context. The additional metadata provides the context to allow an application to perform operations like reference counting, object references, etc.
+
+## Counters ##
+Counters in SAI use 2 patterns.
+- Counters are retrieved using the objects OID and call the associated get_stats() API. sai_get_vlan_stats_fn for SAI_OBJECT_TYPE_VLAN is an example.
+- Counters are retrieved by attaching an explicit counter object to an attribute of the object counters are associated with. SAI_ACL_ENTRY_ATTR_ACTION_COUNTER is used to attach a SAI_OBJECT_TYPE_ACL_COUNTER object. The stats are then obtained by calling sai_get_acl_counter_attribute_fn.
+
+TBD: Use one of the above patterns to retrieve counters based on feedback
+
+## CRM ##
+Table availability and other characteristics can be queried similarly to how the rest of the SAI objects work. sai_object_type_get_availability() allows passing sai_json_t type with the object_name encoded in the attribute list.
```

Comment:
We do not have a sai_object_type per generic object name. i.e. we do not have a unique object type for vip_table objects.  So attribute based filtering is not available in the CRM APIs for these Generic Programmable Entries.

This constraints the functioning of several existing SAI capability query and CRM APIs. 
For example, how will the below work?

- sai_query_attribute_capability()
- sai_query_attribute_enum_values_capability()